### PR TITLE
Update Connector-CRTC bindings after resume in DRM backend

### LINKED ISSUE #204: "Atomic commit failed (modeset): Invalid argument" after switching to Gnome and backward
Reproduced as (assuming Gnome is already running)
1. Start rootston from tty
2. Switch to Gnome
Screen blinks. Looks like Gnome makes clean DRM initialization.
3. Switch back
Gnome interface will be left frozen on screen.

Rootston log says:
```
2017-10-05 07:48:52 - [backend/drm/backend.c:80] DRM fd paused
2017-10-05 07:48:55 - [backend/drm/backend.c:65] DRM fd resumed
2017-10-05 07:48:55 - [backend/drm/atomic.c:57] VGA-1: Atomic commit failed (modeset): Invalid argument
2017-10-05 07:48:55 - [backend/drm/atomic.c:57] HDMI-A-1: Atomic commit failed (modeset): Invalid argument
```

Not reproduced with _weston_ or _xfwm4_. But when switching from Gnome to _weston_ or _xfwm4_ screen also blinks, so maybe they detect such situations and make clean DRM initialization too.

### PR DESCRIPTION
**Pull request by versusvoid**
Fix #204

~~Fix `match_obj()` behaviour on `[0, 0b11], [UNMATCHED, 1]`~~ (not required)
- [x] Fix deactivated monitors check

### CODE REVIEW

**Comment by ascent12** [comment]
I don't like this solution. It only provides a patchwork solution instead of solving the actual core problem.

We shouldn't need to go through the potentially very expensive operation of reconfiguring on VT switch. The old configuration we have should still be good, and we should be able to just keep using it.

I liked the "commit everything at once" solution, but it just needed to be hidden behind the atomic interface properly, and modified to reduce a bunch of code duplication.

**Comment by versusvoid** [comment]
"commit everything at once" means "I don't care what configuration you have, I want you to reconfigure to exact this state"
What is done here is "Get current configuration and try to change it as little as possible".

I see room for improvements in splitting `wlr_drm_connector_set_mode()` and calling `wlr_drm_connector_start_renderer()` only once per connector.

**Comment by SirCmpwn** [comment]
@ascent12 any new comments?

File: backend/drm/backend.c [hunk 1]
```diff
@@ -57,16 +57,69 @@ bool wlr_backend_is_drm(struct wlr_backend *b) {
 	return b->impl == &backend_impl;
 }
 
+static void wlr_drm_connector_update_crtc_binding(struct wlr_drm_connector *conn,
```
Comment (ascent12) [inline backend/drm/backend.c]: This type of function belongs in drm.c.

File: backend/drm/backend.c [hunk 2]
```diff
@@ -57,16 +57,69 @@ bool wlr_backend_is_drm(struct wlr_backend *b) {
...
+		struct wlr_drm_backend *drm) {
+	drmModeConnectorPtr drm_conn = drmModeGetConnector(drm->fd, conn->id);
```
Comment (ascent12) [inline backend/drm/backend.c]: `drmModeConnector *drm_conn`
I really don't like typedefs on pointers.

File: backend/drm/backend.c [hunk 3]
```diff
@@ -57,16 +57,69 @@ bool wlr_backend_is_drm(struct wlr_backend *b) {
...
+	if (!drm_conn) {
+		wlr_log_errno(L_ERROR, "Failed to get DRM connector %u", conn->id);
+		return;
+	}
+	if (!drm_conn->encoder_id) {
+		conn->crtc = NULL;
+		drmModeFreeCrtc(conn->old_crtc);
+		conn->old_crtc = NULL;
+		conn->state = WLR_DRM_CONN_DISCONNECTED;
```
Comment (ascent12) [inline backend/drm/backend.c]: `wlr_drm_connector_cleanup(conn)`

File: backend/drm/backend.c [hunk 4]
```diff
@@ -57,16 +57,69 @@ bool wlr_backend_is_drm(struct wlr_backend *b) {
...
+
+		drmModeFreeConnector(drm_conn);
+		return;
+	}
+
+	drmModeEncoderPtr drm_encoder = drmModeGetEncoder(drm->fd,
```
Comment (ascent12) [inline backend/drm/backend.c]: Same as drm_conn

File: backend/drm/backend.c [hunk 5]
```diff
@@ -57,16 +57,69 @@ bool wlr_backend_is_drm(struct wlr_backend *b) {
...
+				conn->crtc = &drm->crtcs[i];
+			}
+		}
+	}
+
+	drmModeFreeEncoder(drm_encoder);
+	drmModeFreeConnector(drm_conn);
+}
+
+static void wlr_drm_backend_update_connector_crtc_bindings(
```
Comment (ascent12) [inline backend/drm/backend.c]: I wouldn't bother with a function this small, especially if it's only going to be used in one place.
I would just write the for loop where it's going to be used.

File: backend/drm/drm.c [hunk 6]
```diff
@@ -416,7 +424,8 @@ static bool wlr_drm_connector_set_mode(struct wlr_output *output,
 		goto error_conn;
 	}
 
-	realloc_crtcs(drm, conn);
+	memset(changed_outputs, false, sizeof(bool) * drm->outputs->length);
```
Comment (ascent12) [inline backend/drm/drm.c]: `sizeof(changed_outputs)`

**Comment by ascent12** [comment]
This is going to mess with #290, so I'm going to wait for that to be finished and have the list_t related changes in this PR before I would merge it.

File: backend/drm/backend.c (hunk 2 above)
Comment (SirCmpwn) [reply backend/drm/backend.c]: typedefs are not allowed at all in our codebase.

Comment (ascent12) [reply backend/drm/backend.c]: It's not our typedef, but they provide both a struct and a pointer to struct typedef in libdrm for some stupid reason.

**Comment by versusvoid** [comment]
`wlr_drm_connector_cleanup()` release surfaces, planes and emits `output_remove` signal. None of it is required here, so I'm not calling it.

**Comment by versusvoid** [comment]
I got rid of new function and set `crtc` field in `wlr_drm_scan_connectors()`. It's effectively the same but with less code.
Current CRTCs are always considered when matching with connectors now. This is workaround _unoptimal_ result of `match_obj()`.

**Pull request merged**